account_invoice_mode_monthly/models/sale_order.py

```diff
@@ -0,0 +1,100 @@
+            ],
+            ["partner_invoice_id"],
+            groupby=["partner_invoice_id"],
+        )
+        for partner in partner_ids:
+            self.with_delay()._generate_invoices_by_partner(
+                partner["partner_invoice_id"][0]
+            )
+        companies.write({"invoicing_mode_monthly_last_execution": datetime.now()})
+        return partner_ids
+
+    @job(default_channel="root.invoice_monthly")
+    def _generate_invoices_by_partner(self, partner_id, invoicing_mode="monthly"):
+        """Generate invoices for a customer sales order."""
+        partner = self.env["res.partner"].browse(partner_id)
+        if partner.invoicing_mode != invoicing_mode:
+            return "Customer {} is not configured for monthly invoicing.".format(
+                partner.name
+            )
+        sales = self.search(
+            [
+                ("invoice_status", "=", "to invoice"),
+                ("partner_invoice_id", "=", partner.id),
+                ("order_line.qty_to_invoice", ">", 0),
+            ]
+        )
+        # By default grouped by partner/currency. Refund are not generated
+        invoices = sales._create_invoices(grouped=partner.one_invoice_per_order)
+        for invoice in invoices:
+            invoice.with_delay()._validate_invoice()
+        return invoices
+
+    @api.model
+    def _company_monthly_invoicing_today(self):
+        """Get company ids for which today is monthly invoicing day."""
+        today = datetime.now()
+        first_day_this_month = today.replace(
+            day=1, hour=0, minute=0, second=0, microsecond=0
+        )
+        first_day_last_month = first_day_this_month - relativedelta(months=1)
+        last_day_of_this_month = (today + relativedelta(day=31)).day
+        # Last month still not executed, it needs to be done
+        domain_last_month = [
+            ("invoicing_mode_monthly_last_execution", "<", first_day_last_month),
+        ]
+        # Invoicing day is today or in the past and invoicing not yet done
+        domain_this_month = [
+            "|",
+            ("invoicing_mode_monthly_last_execution", "<", first_day_this_month),
+            ("invoicing_mode_monthly_last_execution", "=", False),
+            ("invoicing_mode_monthly_day_todo", "<=", today.day),
+        ]
+        # Make sure non exisiting days are done at the end of the month
+        domain_last_day_of_month = [
+            "|",
+            ("invoicing_mode_monthly_last_execution", "<", first_day_this_month),
+            ("invoicing_mode_monthly_last_execution", "=", False),
+            ("invoicing_mode_monthly_day_todo", ">", today.day),
+        ]
+        if today.day == last_day_of_this_month:
+            domain = OR(
+                [domain_last_month, domain_this_month, domain_last_day_of_month]
+            )
+        else:
+            domain = OR([domain_last_month, domain_this_month])
+
+        return self.env["res.company"].search(domain)
```

Comment:
I feel like that method is kind of complicated for what it does.
Would it be easier to compute the next invoicing day at the end of the invoicing, and return companies where `("next_invoice_day", "=", "fields.date.today()")` ?

Reply:
Yes, then for any company with `("next_invoice_day", "<=", "field.date.today()")`, that method should be run.
You just have to change the operator. Once that method is run on a record, you can update the date on it.
Except for the special case you mentioned, you shouldn't have some `next_invoice_day` inferior to today, am I right ?

Reply:
I probably don't have the full picture of it, though.

Reply:
Also, it may be complicated to compute the next month given a day.

Reply:
@TDu ?

Reply:
I think this has been answered by this comment https://github.com/OCA/account-invoicing/pull/755#discussion_r625023369